Mhoffm promremote performance improvements

mhoffm-aiven opened the pull request:
# Required for all PRs

<!-- Before opening a pull request you should run the following checks locally to make sure the CI will pass.

make lint
make check
make check-deps
make test
make docs

-->

<!-- Complete the tasks in the following list. Change [ ] to [x] to
show completion. -->

- [ ] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [ ] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

<!-- Link to issues that describe the need for the change. Issues
should include context that will help reviewers understand why the
change is needed.

Make sure to link issues and using a keyword like "resolves #1234".
https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

resolves #

<!-- Finally, include a summary of the code change itself. This
description should tell reviewers how the issues were resolved.

example: Fixed an off by one error in counter variable in type FooBar.

example: Added an input plugin to gather yak shaving metrics using
golang library yaktech/shaver. -->

powersj commented:
Hi @mhoffm-aiven,

Thanks for the PR. Is there an issue that goes along with this? If not could you file one please?

Also looks like there are a few lint issues:

```
  plugins/serializers/prometheus/convert.go:58:6                                 unused    func `isValid` is unused
  plugins/serializers/prometheusremotewrite/prometheusremotewrite.go:56:80       revive    empty-lines: extra empty line at the end of a block
  plugins/serializers/prometheusremotewrite/prometheusremotewrite_test.go:37:3   revive    unhandled-error: Unhandled error in call to function s.SerializeBatch
  plugins/serializers/prometheusremotewrite/prometheusremotewrite_test.go:37:19  errcheck  Error return value of `s.SerializeBatch` is not checked
```

Thanks!

mhoffm-aiven commented:
> Hi @mhoffm-aiven,
> 
> Thanks for the PR. Is there an issue that goes along with this? If not could you file one please?
> 
> Also looks like there are a few lint issues:
> 
> ```
>   plugins/serializers/prometheus/convert.go:58:6                                 unused    func `isValid` is unused
>   plugins/serializers/prometheusremotewrite/prometheusremotewrite.go:56:80       revive    empty-lines: extra empty line at the end of a block
>   plugins/serializers/prometheusremotewrite/prometheusremotewrite_test.go:37:3   revive    unhandled-error: Unhandled error in call to function s.SerializeBatch
>   plugins/serializers/prometheusremotewrite/prometheusremotewrite_test.go:37:19  errcheck  Error return value of `s.SerializeBatch` is not checked
> ```
> 
> Thanks!

Hey @powersj,

No issue yet ( ill create one in a minute ); i was investigating performance issues with our telegraf promremote output and did some profiling.

powersj commented:
>  i was investigating performance issues with our telegraf promremote output and did some profiling.

ok thanks - if you could include some of that data I would appreciate it! thanks again!

mhoffm-aiven commented:
> > i was investigating performance issues with our telegraf promremote output and did some profiling.
> 
> ok thanks - if you could include some of that data I would appreciate it! thanks again!

I run that benchmark form this PR against main and got locally 
```
 fedora  ~  git  …  plugins  serializers  prometheusremotewrite  master 
$ go test -run=^$ -bench=. -benchtime 30s -benchmem -memprofile memprofile.out -cpuprofile profile.out
goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkRemoteWrite-8   	   38106	    924518 ns/op	  648409 B/op	    7005 allocs/op
PASS
ok  	github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite	44.943s
```
against this PR:
```
$ go test -run=^$ -bench=. -benchtime 30s -benchmem -memprofile memprofile.out -cpuprofile profile.out
goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkRemoteWrite-8   	   64338	    539615 ns/op	  344826 B/op	    4007 allocs/op
PASS
ok  	github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite	40.534s
```

main offenders were copied slices ( i fixed this by allocating once and truncating on loop iterations ) and sort.Slice instead of sort.Sort inplace. surprisingly `unicode.Is` also was pretty high up in the profile. Let me run against both branches and upload profiles a bit later on.

mhoffm-aiven commented:
top allocs with the benchmark on master (notice `reflectlite.Swapper` and `createLabels`:
```
Type: alloc_space
Time: Mar 29, 2023 at 10:37am (CEST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 28871.76MB, 99.93% of 28890.89MB total
Dropped 48 nodes (cum <= 144.45MB)
      flat  flat%   sum%        cum   cum%
14601.41MB 50.54% 50.54% 19582.68MB 67.78%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.getPromTS
 8599.07MB 29.76% 80.30%  8599.07MB 29.76%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.(*Serializer).createLabels
 4981.27MB 17.24% 97.55%  4981.27MB 17.24%  internal/reflectlite.Swapper
  688.01MB  2.38% 99.93%   688.01MB  2.38%  github.com/influxdata/telegraf/plugins/serializers/prometheus.MetricName
       2MB 0.0069% 99.93% 28883.26MB   100%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.(*Serializer).SerializeBatch
         0     0% 99.93% 28883.26MB   100%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.BenchmarkRemoteWrite
         0     0% 99.93%  4981.27MB 17.24%  sort.Slice
         0     0% 99.93% 28882.76MB   100%  testing.(*B).launch
         0     0% 99.93% 28883.26MB   100%  testing.(*B).runN
```
top 10 on this PR:
```
Type: alloc_space
Time: Mar 29, 2023 at 10:40am (CEST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 23.15GB, 99.74% of 23.21GB total
Dropped 52 nodes (cum <= 0.12GB)
      flat  flat%   sum%        cum   cum%
   22.02GB 94.87% 94.87%    22.02GB 94.87%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.getPromTS
    1.12GB  4.81% 99.68%     1.12GB  4.81%  github.com/influxdata/telegraf/plugins/serializers/prometheus.MetricName
    0.01GB 0.055% 99.74%    23.20GB 99.95%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.(*Serializer).SerializeBatch
         0     0% 99.74%    23.20GB 99.95%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.BenchmarkRemoteWrite
         0     0% 99.74%    23.20GB 99.95%  testing.(*B).launch
         0     0% 99.74%    23.20GB 99.95%  testing.(*B).runN
```

cpu profile on master:
```
Type: cpu
Time: Mar 29, 2023 at 10:37am (CEST)
Duration: 44.63s, Total samples = 63.74s (142.82%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 27.73s, 43.50% of 63.74s total
Dropped 268 nodes (cum <= 0.32s)
Showing top 10 nodes out of 114
      flat  flat%   sum%        cum   cum%
     4.77s  7.48%  7.48%     17.97s 28.19%  runtime.mallocgc
     3.43s  5.38% 12.86%      4.05s  6.35%  runtime.heapBitsSetType
     3.39s  5.32% 18.18%      3.40s  5.33%  hash/fnv.(*sum64a).Write
     3.21s  5.04% 23.22%     10.36s 16.25%  runtime.scanobject
     2.52s  3.95% 27.17%      2.52s  3.95%  runtime.nextFreeFast (inline)
     2.33s  3.66% 30.83%      4.40s  6.90%  unicode.Is
     2.07s  3.25% 34.08%      2.07s  3.25%  unicode.is16
     2.04s  3.20% 37.28%      8.71s 13.66%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.(*Serializer).createLabels
     2.03s  3.18% 40.46%      7.27s 11.41%  github.com/influxdata/telegraf/plugins/serializers/prometheus.isValid
     1.94s  3.04% 43.50%      1.94s  3.04%  runtime.memmove
```
and on this PR:
```
Type: cpu
Time: Mar 29, 2023 at 10:40am (CEST)
Duration: 40.91s, Total samples = 56.60s (138.36%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top
Showing nodes accounting for 27.17s, 48.00% of 56.60s total
Dropped 227 nodes (cum <= 0.28s)
Showing top 10 nodes out of 106
      flat  flat%   sum%        cum   cum%
     5.44s  9.61%  9.61%      5.44s  9.61%  hash/fnv.(*sum64a).Write
     4.14s  7.31% 16.93%     15.25s 26.94%  runtime.mallocgc
     3.57s  6.31% 23.23%      4.19s  7.40%  runtime.heapBitsSetType
     2.86s  5.05% 28.29%      3.64s  6.43%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.(*Serializer).appendCommonLabels
     2.51s  4.43% 32.72%      7.88s 13.92%  runtime.scanobject
     2.01s  3.55% 36.27%      2.01s  3.55%  runtime.nextFreeFast (inline)
     1.91s  3.37% 39.65%      1.91s  3.37%  runtime.memmove
     1.74s  3.07% 42.72%      1.78s  3.14%  runtime.pageIndexOf (inline)
     1.52s  2.69% 45.41%     42.41s 74.93%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.(*Serializer).SerializeBatch
     1.47s  2.60% 48.00%      1.47s  2.60%  runtime.futex
```

powersj commented on plugins/serializers/prometheus/convert.go:
```diff
@@ -55,35 +55,30 @@ var LabelNameTable = Table{
 }
 
-func isValid(name string, table Table) bool {
-	if name == "" {
-		return false
```
as we return true by default, did we lose the check for an empty sting with this? or was it checked elsewhere?

mhoffm-aiven replied:
yeah this would pass:
```
		{
			name: "empty name",
			metric: testutil.MustMetric(
				"",
				map[string]string{
					"host": "example.org",
				},
				map[string]interface{}{
					"time_idle": 42.0,
				},
				time.Unix(0, 0),
			),
			expected: []byte(`
_time_idle{host="example.org"} 42
`),

```

mhoffm-aiven replied:
I dug a bit and this would actually produce the same result on main since the only way to produce an empty metric name would be if the metric is called "prometheus". For empty tag this was a regression though, i added some tests and fixed your comments! thank you, i missed this

powersj commented on plugins/serializers/prometheus/convert.go:
```diff
@@ -55,35 +55,30 @@ var LabelNameTable = Table{
+func isValidMetricName(name string) bool {
+	for i := 0; i < len(name); i++ {
+		b := name[i]
+		if !((b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || b == ':' || (b >= '0' && b <= '9' && i > 0)) {
+			return false
+		}
```
out of curiosity did you compare this to a regex comparison e.g. `[a-zA-Z_:][a-zA-Z0-9_:]*`?

mhoffm-aiven replied:
No i did not compare; i can try in a bit though

mhoffm-aiven replied:
its much more expensive actually!

```
$ go test -run=^$ -bench=. -benchtime 30s -benchmem -memprofile memprofile.out -cpuprofile profile.out
goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkRemoteWrite-8   	   38215	    931852 ns/op	  350204 B/op	    4008 allocs/op
PASS
ok  	github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite	45.243s
```

```
(pprof) top
Showing nodes accounting for 26.77s, 49.44% of 54.15s total
Dropped 240 nodes (cum <= 0.27s)
Showing top 10 nodes out of 98
      flat  flat%   sum%        cum   cum%
     5.75s 10.62% 10.62%      5.75s 10.62%  regexp.(*bitState).shouldVisit (inline)
     5.54s 10.23% 20.85%     16.06s 29.66%  regexp.(*Regexp).tryBacktrack
     2.77s  5.12% 25.96%      2.77s  5.12%  hash/fnv.(*sum64a).Write
     2.76s  5.10% 31.06%      2.76s  5.10%  regexp/syntax.(*Inst).MatchRunePos
     2.45s  4.52% 35.59%      8.39s 15.49%  runtime.mallocgc
     1.82s  3.36% 38.95%      2.09s  3.86%  runtime.heapBitsSetType
     1.78s  3.29% 42.23%     14.14s 26.11%  github.com/influxdata/telegraf/plugins/serializers/prometheusremotewrite.(*Serializer).appendCommonLabels
     1.48s  2.73% 44.97%      1.48s  2.73%  runtime.memmove
     1.33s  2.46% 47.42%      4.14s  7.65%  runtime.scanobject
     1.09s  2.01% 49.44%      1.09s  2.01%  runtime.memclrNoHeapPointers

```

powersj replied:
ok thanks for trying it out!

powersj reviewed (commented):
Thanks for the PR. Have you been running this successfully? Couple questions in line.

MichaHoffmann commented:
> Thanks for the PR. Have you been running this successfully? Couple questions in line.

No this has not been running in our production yet.

powersj reviewed (approved):
Thanks!

srebhan commented on plugins/serializers/prometheus/convert.go:
```diff
@@ -55,35 +55,36 @@ var LabelNameTable = Table{
-			if !unicode.In(r, table.Rest) {
-				return false
-			}
+	for i := 0; i < len(name); i++ {
+		b := name[i]
+		if !((b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || b == ':' || (b >= '0' && b <= '9' && i > 0)) {
+			return false
 		}
 	}
```
I really would love to see a (pre-compiled) regexp for this. Is there any reason for doing this manual dance?

srebhan commented on plugins/serializers/prometheus/convert.go:
```diff
@@ -55,35 +55,36 @@ var LabelNameTable = Table{
+	return true
+}
 
+func isValidLabelName(name string) bool {
+	if len(name) == 0 {
+		return false
+	}
+	for i := 0; i < len(name); i++ {
+		b := name[i]
+		if !((b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0)) {
+			return false
+		}
+	}
 	return true
 }
```
Same as above.

mhoffm-aiven replied:
Same as above

srebhan commented on plugins/serializers/prometheusremotewrite/prometheusremotewrite.go:
```diff
@@ -52,10 +54,12 @@ func (s *Serializer) Serialize(metric telegraf.Metric) ([]byte, error) {
 
 func (s *Serializer) SerializeBatch(metrics []telegraf.Metric) ([]byte, error) {
-	var buf bytes.Buffer
+	s.buf.Reset()
```
This function now becomes stateful and will do strange things if called concurrently. Can we please at least add a mutex to protect the buffer?

mhoffm-aiven replied:
went back to the stateless approach; it does not have big performance impacts and state is not needed.

srebhan commented on plugins/serializers/prometheusremotewrite/prometheusremotewrite.go:
```diff
@@ -239,8 +236,10 @@ func (s *Serializer) SerializeBatch(metrics []telegraf.Metric) ([]byte, error) {
 	encoded := snappy.Encode(nil, data)
-	buf.Write(encoded) //nolint:revive // from buffer.go: "err is always nil"
-	return buf.Bytes(), nil
+	s.buf.Write(encoded) //nolint:revive // from buffer.go: "err is always nil"
+
+	out := make([]byte, 0, s.buf.Len())
+	return append(out, s.buf.Bytes()...), nil
```
Hmmm you are doing the alloc here with the length of the plugin-global buffer but otherwise scratch on that buffer. How about only storing the _maximal-allocated-buffer-size_ in the plugin struct and alloc that locally. Wouldn't that yield the same improvement as line 241 will also alloc once?

Even better, you will not introduce a state (that requires locking) and you don't need to copy as you can directly write to the allocated `out` buffer...

srebhan reviewed (changes requested):
That's an awesome improvement @mhoffm-aiven! Thank you for this! I do have one comment though as now the batch-serialization will break if called concurrently. Please also find a suggestion inline...

srebhan commented:
@powersj can you please take a look again after the changes!?
